### DISCUSSION TITLE
How to output the grid metrics?

### DISCUSSION THREAD

**Discussion by NoraLoose** [CliMA/Oceananigans.jl#3986] [General]
Is there a way to write the grid metrics (distances, areas, volumes) to file?

**Comment by glwagner** [comment]
@tomchor and @ali-ramadhan may be able to answer this for `NetCDFOutputWriter`. I think some metrics are saved but not all because there's been talk of improving the situation.

For `JLD2OutputWriter` you might want to write an `init` function. However, if you're staying in julia it is simpler to just rebuild the grid, eg writing something like

```julia
ut = FieldTimeSeries(filename, "u") # assumes you saved a field called "u"
grid = ut.grid
Δx = xspacings(grid, Face(), Center(), Center())
```

I think there is an issue somewhere about compatibility with xarray / xgcm if that is specifically what you are looking for.

**Reply by glwagner** [reply]
Here's a few: https://github.com/CliMA/Oceananigans.jl/issues/2248, https://github.com/CliMA/Oceananigans.jl/issues/1334

**Reply by glwagner** [reply]
Very recently we overhauled the user-facing julia abstraction for representing spacings, which I think was one of the preliminary pieces of work needed before getting into this.

**Reply by NoraLoose** [reply]
Thanks @glwagner! Does the code snippet above work for `filename` being a checkpoint file (written via `Checkpointer()`)?

**Reply by glwagner** [reply]
I believe the grid is serialized in the Checkpointer but you may have to open the file with `JLD2` to get it. I can figure that out.

**Reply by glwagner** [reply]
This should work:

```julia
julia> using JLD2, Oceananigans

julia> filename = "north_atlantic_calibration_1.0_i0_e0_iteration97200.jld2"
"north_atlantic_calibration_1.0_i0_e0_iteration97200.jld2"

julia> file = jldopen(filename)
JLDFile /Users/gregorywagner/Projects/ClimaOceanCalibration.jl/test/north_atlantic_calibration_1.0_i0_e0_iteration97200.jld2 (read-only)
 ├─🔢 checkpointed_properties
 ├─🔢 grid
 ├─🔢 particles
 ├─🔢 clock
 ├─🔢 coriolis
 ├─🔢 buoyancy
 ├─🔢 closure
 ├─📂 timestepper
 │  └─ ⋯ (2 more entries)
 └─ ⋯ (9 more entries)

julia> grid = file["grid"]
100×60×60 ImmersedBoundaryGrid{Float64, Bounded, Bounded, Bounded} on CPU with 7×7×3 halo:
├── immersed_boundary: GridFittedBottom(mean(z)=-2259.77, min(z)=-6000.0, max(z)=0.0)
├── underlying_grid: 100×60×60 LatitudeLongitudeGrid{Float64, Bounded, Bounded, Bounded} on CPU with 7×7×3 halo and with precomputed metrics
├── longitude: Bounded  λ ∈ [-100.0, 0.0]  regularly spaced with Δλ=1.0
├── latitude:  Bounded  φ ∈ [0.0, 60.0]    regularly spaced with Δφ=1.0
└── z:         Bounded  z ∈ [-6000.0, 0.0] variably spaced with min(Δz)=5.24961, max(Δz)=438.409

julia> close(file)
```

A key caveat is that the version of Oceananigans may need to be the same. Specifically grids were _very recently_ updated between 0.94 and 0.95 to support z* vertical coordinate. So you won't be able to deserialize a grid that was saved pre-0.94 if your project is updated to version 0.95.

**Reply by glwagner** [reply]
The grids do not change very often otherwise...

**Reply by NoraLoose** [reply]
> For JLD2OutputWriter you might want to write an init function. However, if you're staying in julia it is simpler to just rebuild the grid, eg writing something like
> 
> ```
> ut = FieldTimeSeries(filename, "u") # assumes you saved a field called "u"
> grid = ut.grid
> Δx = xspacings(grid, Face(), Center(), Center())
> ```

I guess there is no easy equivalent for getting at the grid **area**, is there? Specifically, I have a `LatitudeLongitudeGrid`, which I rebuild from the pickup file as you describe above, but then I would still like to convert
```
grid.Azᶠᶠᵃ
```
to a `Field{Center, Center, Center}(grid)` field.

**Reply by simone-silvestri** [reply]
We don't have those utilities yer at the moment. 
You can still build the metrics using a `KernelFunctionOperation` as such:
```julia
ut = FieldTimeSeries(filename, "u") # assumes you saved a field called "u"
grid = ut.grid
Az = Field(KernelFunctionOperation{Face, Face, Center}(zarea, grid, Face(), Face(), Center()))
compute!(Az)
```

**Comment by tomchor** [comment]
Like @glwagner mentioned, we don't currently output all the grid metrics to NeTCDF, but @ali-ramadhan and I are planning on dedicating some time to implementing this soon-ish. We've been finding it hard to find the time though, se let me know if you're interested in helping out :)

In the meantime, this is what I've been personally using whenever I need xgcm integration: https://pastebin.com/vSbxz7mU
I don't remember if I adapted this to https://github.com/CliMA/Oceananigans.jl/pull/3974 (if any adaptation is needed), so it might need some small tweaks.

**Reply by glwagner** [reply]
What is all of this code doing that's needed for integration with xgcm?

**Reply by tomchor** [reply]
Sorry, I should have been more specific. It outputs spacings for grids following [`xgcm`'s standards](https://xgcm.readthedocs.io/en/latest/grids.html).

**Reply by NoraLoose** [reply]
Thanks @tomchor for the update and the link to your code - very helpful! I will try to help integrate this into the source code.

**Reply by glwagner** [reply]
Is the reduction of the spacing by 1/2 on the boundaries required for xgcm?

**Reply by NoraLoose** [reply]
> Is the reduction of the spacing by 1/2 on the boundaries required for xgcm?

No, that seems odd to me. Do you refer to dividing for example the x-direction at u-points by 2, if the u-point lies on the boundary between land and ocean? In that case, the value at that u-point should probably be masked anyway. This reminds me that we should provide all the masks as part of the grid file as well.

**Reply by glwagner** [reply]
I think that's what this code goes for example:

```julia
boundary_xspacing(i,j,k,grid, ℓx, ℓy, ℓz) = ifelse(i==1, xnode(i,j,k, grid, flip_loc(ℓx), ℓy, ℓz) - xnode(i,j,k, grid, ℓx, ℓy, ℓz),
                                                         xnode(i,j,k, grid, ℓx, ℓy, ℓz)           - xnode(i-1,j,k, grid, flip_loc(ℓx), ℓy, ℓz))
```

not on immersed boundaries here but just at the ends of the domain (presumably this only makes sense for Bounded directions).

**Reply by glwagner** [reply]
We can generate mask using the operator `inactive_cell`, or `immersed_cell` if only for immersed boundaries.

**Reply by glwagner** [reply]
What does MOM6 do about the fact that the grid / vertical coordinate is moving around?

**Reply by tomchor** [reply]
> Is the reduction of the spacing by 1/2 on the boundaries required for xgcm?

afaik that's required for cell faces on bounded directions. The idea is that Δx at a point is the spacing (in terms of wet space) _around_ that point.

If that point is a cell center, then the spacing is always gonna be the distance between the cell faces to the left and right of that point. The same is true for faces away from boundaries. But for faces at boundaries, one of the sides around the face is not "wet" (meaning it doesn't hold any of the fluid being modeled), be it because it's outside of the domain interior or because it's adjacent to an immersed boundary. That way we always get `sum(grid.Δxᶠᵃᵃ) == Lx` and `sum(field * grid.Δxᶠᵃᵃ) / sum(grid.Δxᶠᵃᵃ) == Average(field, dims=1)`.

In our case, for rectilinear grids, I think that translates to cutting the nominal spacing at the boundaries in half (except maybe when the grid is stretched?).

Admittedly this is a bit confusing, and it was only clear to me after a few discussions with the `xgcm` people. I'll try to link the discussions here if I can find them.

Let me know if this isn't clear or if I missed something.

**Reply by NoraLoose** [reply]
I understand your point @tomchor. But if only half of the cell in question is wet and the other half is dry, then this point is most likely masked in the underlying fluid model, and the value at that point would not enter the average that you compute in your example. 

So both solutions (leaving boundary distances as they are and dividing boundary distances by two) will lead to the same result in your computation. This is why I would go with the simpler (less confusing) way to leave the boundary distances untouched. I think this problem should rather be addressed by providing the masks.

**Reply by NoraLoose** [reply]
So your computation could be
```
sum(mask * field * grid.Δxᶠᵃᵃ) / sum(mask * grid.Δxᶠᵃᵃ) == Average(field, dims=1)
```

**Reply by NoraLoose** [reply]
> What does MOM6 do about the fact that the grid / vertical coordinate is moving around?

`h` (the layer thickness) is a prognostic time-dependent variable that you can (and should) output alongside the other prognostic variables `u`, `v`, ...

**Reply by NoraLoose** [reply]
Are you asking because of z-star vertical coordinates?

**Reply by NoraLoose** [reply]
I think about z-star as a different case. For z-star, we could provide a helper to compute the vertical coordinate spacing from the reference levels and the SSH. Otherwise, that's a lot of output?

**Reply by NoraLoose** [reply]
How did we manage to entangle so much stuff in this single thread?

**Reply by glwagner** [reply]
It's a great thread!

**Reply by glwagner** [reply]
> I think about z-star as a different case. For z-star, we could provide a helper to compute the vertical coordinate spacing from the reference levels and the SSH. Otherwise, that's a lot of output?

It's just that the grid information is incomplete without information about the metrics that fluctuate. We were discussing automatically adding the stretching factors with any output so that the correct grid at any moment in time can be constructed. This is relevant for saving down the grid metrics since I guess in the case of `ZStarVerticalCoordinate` we will be saving the reference grid without stretching factors. I guess I'm wondering how this works re: xgcm and if there are any special considerations.

**Reply by NoraLoose** [reply]
xgcm can build a grid from a time-dependent vertical coordinate as, e.g., seen in [this MOM6 example](https://mom6-analysiscookbook.readthedocs.io/en/latest/notebooks/Buoyancy_Geostrophic_shear.html#).

**Reply by NoraLoose** [reply]
In that example, the user told MOM6 to output the variable `volcello`, which is the time-dependent volume of each grid cell.

**Reply by glwagner** [reply]
Nice, starting to understand! It looks like xgcm knows magically what `volcello` means too? So that is the name we should use for this application...

**Reply by NoraLoose** [reply]
No, `xgcm` does not look for the name `volcello`. Usually, you tell `xgcm` what the names of your grid metrics are, see cell 4 in [this notebook](https://xgcm.readthedocs.io/en/latest/grid_metrics.html). There are some checks within `xgcm` to automatically figure out which of the variables in your dataset are the metrics, so this is how it might gotten to `volcello`. But there is nothing special about this name, so we don't need to adopt it (luckily)!

**Reply by glwagner** [reply]
I was trying to figure out if `volcello` is some kind of standard name defined by CF or S-grid but it doesn't look like it is.

**Reply by NoraLoose** [reply]
Haha, looking back at the mysterious [`volcello`-notebook](https://mom6-analysiscookbook.readthedocs.io/en/latest/notebooks/Buoyancy_Geostrophic_shear.html#) it seems to me that there is no computation in that notebook that actually needs the grid cell volume. I don't think `volcello` is part of the Grid object that is created in cell 4.

**Reply by tomchor** [reply]
> I understand your point @tomchor. But if only half of the cell in question is wet and the other half is dry, then this point is most likely masked in the underlying fluid model, and the value at that point would not enter the average that you compute in your example.
> 
> So both solutions (leaving boundary distances as they are and dividing boundary distances by two) will lead to the same result in your computation. This is why I would go with the simpler (less confusing) way to leave the boundary distances untouched. I think this problem should rather be addressed by providing the masks.

I see your point. So the mask would be 1 in all center points, half at the edges (for face fields in a bounded topology) and zero outside the wet domain?

**Reply by glwagner** [reply]
Plus on the staggered grid I don't think it's quite right to think of the "peripheral cells" (as we have termed them in Oceananigans) as "half wet, half dry". Instead these cells either have boundary conditions (like the boundary normal velocities), or have to be diagnosed from other quantities correctly by applying the boundary conditions on those other quantities (for example the gradient of a field normal to a boundary). It's not correct to divide the metrics by two, I don't think, either (in a finite volume sense).

In other words the C-grid provides a set of rules that we can apply to understand the behavior of the peripheral points, but in no situation is it correct to think of the peripheral volumes as having half of the full cell volume.
